File: ch16/yojkim/ch16/money.go
package ch16

type Money struct {
	currency string
	amount int
}

func NewMoney(amount int, currency string) Money {
	return Money{
		amount: amount,
		currency: currency,
	}
}

func NewDollar(amount int) Money {
	return NewMoney(amount, "USD")
}

func NewFranc(amount int) Money {
	return NewMoney(amount, "CHF")
}

func (m Money) Times(multiplier int) Expression {
	return NewMoney(m.amount * multiplier, m.currency)
}

func (m Money) Equal(input Money) bool {
	if m.currency != input.currency {
		return false
	}
	return m.amount == input.amount
}

func (m Money) Currency() string {
	return m.currency
}

func (m Money) Plus(addend Expression) Expression {
	return NewSum(m, addend)
}

func (m Money) Reduce(bank Bank, to string) Money {
	rate := bank.Rate(m.currency, to)
	return NewMoney(m.amount / rate, to)
}

File: ch15/gunddol/src/ch15/Money.java
package ch15;

class Money implements Expression{

    protected String currency;
    protected int amount;

    Expression times(int multiplier){
        return new Money(amount * multiplier, currency);
    }

    Money(int amount, String currency){
        this.amount = amount;
        this.currency = currency;
    }

    public static Money dollar(int amount){
        return new Money(amount, "USD");
    }

    public static Money franc(int amount){
        return new Money(amount,"CHF");
    }

    public String toString(){
        return amount + " " + currency;
    }

    String currency(){
        return currency;
    }

    public boolean equals(Object object){
        Money money = (Money) object;
        return amount == money.amount
                && currency().equals(money.currency());
    }

    public Expression plus(Expression addend) {
        return new Sum(this, addend);
    }

    public Money reduce(Bank bank, String to){
        int rate = bank.rate(currency, to);
        return new Money(amount / rate, to);
    }
}
